wallsAndGates.js

```diff
@@ -0,0 +1,141 @@
+/*
+Just as the previous question, we're going to represent our input as a grid where 
+we're only concerned about the the empty rooms (represnted as 2147483647), since we're trying 
+to find the distance for each empty room that is the closest to the nearest gate (represnted as 0). 
+
+For this approach, we're going to use a multi-sourced BFS again to find the distance to the nearest gate 
+and the way that works is we're going to gather the location of all the gates first and add them to the queue,
+that way we can process them together. During the iteration, we're going to BFS each of those 
+locations in all 4 directions and until we reach an empty room. Once we reach an empty room,
+we mark it with the current traveled distance so far.
```

Comment:
Why can we mark it with the current traveled distance so far? Like why is that the correct answer? Or if it's not, how do we know that we will later overwrite this value with the correct answer?

Reply:
Since we're starting from the source (in this context the gate), our starting path will be 0. Once we BFS from the source, we're traversing by levels and we only increase the path by 1 if it's the fist time we're visiting an empty room.

Reply:
The key is that because of the way we use our visited set, we only process each node the very FIRST time that we see it. Because a BFS goes level-by-level, the first time we see a node, we have taken the shortest path to reach that node (if there was a shorter path, we must have seen it first).

Reply:
Yes, I forgot to mention that the visited set is the reason for visiting a location for the first time. And when we're processing a node, it will be for the first time visiting that node and the shortest path.

Reply:
Good.